### ISSUE TITLE
[MSKINS-245] Maven Site 4 will break code highlight of site generated by Markdown

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-fluido-skin#259]
**[Xavi Lee](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER284203)** opened **[MSKINS-245](https://issues.apache.org/jira/browse/MSKINS-245?redirect=false)** and commented

repro repo https://github.com/awxiaoxian2020/code-render-bug
master branch is Maven Site 3 with Fluido skin 1
v4 branch is Maven Site 4 with Fluido skin 2.

Open their respective `target/site/test.html` files in local to see the rendered result.

---

**Attachments:**
- [maven-site-3.png](https://issues.apache.org/jira/secure/attachment/13066435/maven-site-3.png) (_34.40 kB_)
- [maven-site-4.png](https://issues.apache.org/jira/secure/attachment/13066434/maven-site-4.png) (_39.61 kB_)
- [test-v3.html](https://issues.apache.org/jira/secure/attachment/13067367/test-v3.html) (_2.28 kB_)
- [test-v4.html](https://issues.apache.org/jira/secure/attachment/13067366/test-v4.html) (_2.48 kB_)

**Issue Links:**
- [MSKINS-249](https://issues.apache.org/jira/browse/MSKINS-249) Fix for MSKINS-245 does not cover all cases
 (_**"causes"**_)
- [DOXIA-685](https://issues.apache.org/jira/browse/DOXIA-685) Replace SinkEventAttributes#BOXED with #SOURCE to clearly separate between regular verbatim and verbatim source (code)
 (_**"is caused by"**_)
- [MSKINS-237](https://issues.apache.org/jira/browse/MSKINS-237) Rework skin for new site model
 (_**"is caused by"**_)
- [MSKINS-258](https://issues.apache.org/jira/browse/MSKINS-258) Markdown: Fenced code without info string incorrectly formatted by Fluido skin

- [MSKINS-242](https://issues.apache.org/jira/browse/MSKINS-242) Replace deprecated prettify JS library

- [MSKINS-244](https://issues.apache.org/jira/browse/MSKINS-244) Get rid of CSS classes "table" and "verbatim"

- [DOXIA-731](https://issues.apache.org/jira/browse/DOXIA-731) Simplify HTML emitted from Sink.verbatim(...)
 (_**"requires"**_)

**Remote Links:**
- [GitHub Pull Request #58
](https://github.com/apache/maven-fluido-skin/pull/58)
- [GitHub PR #58
](https://github.com/apache/maven-fluido-skin/pull/58)

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin), this is a Markdown-related issue: Do you want to take a look at it? I can confirm that the issue is correct.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Jeremy Landis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hazendaz), thanks for sharing. I guess going down version by version might help to tackle the issue.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Attached are the HTML pages generated with v3 ( [^test-v3.html] ) and v4 ( [^test-v4.html] ).
The difference for the code section looks like this:

```
<main id="bodyColumn"  class="span10" >
<div class="source"><pre class="prettyprint"><code class="language-java">int a = 1;
</code></pre></div>
        </main>
```

(v3 = working) vs

```
<main id="bodyColumn" class="span10">
<div class="verbatim">
<pre><code class="language-java">int a = 1;
</code></pre></div>
        </main>
```

(v4 = non-working)

The JS library for code highlighting is https://github.com/apache/maven-fluido-skin/blob/master/src/main/resources/js/prettify.js.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Do you know where this is maintained and documented. This is hard to see from the minified JS included in Fluido....

Update: Seems this is like a discontinued library from Google: https://meta.stackexchange.com/questions/353983/goodbye-prettify-hello-highlight-js-swapping-out-our-syntax-highlighter. We should probably change that as well (tracked now in MSKINS-242). The legacy prettify library requires the class `prettyprint` though in the `pre` tag (https://github.com/googlearchive/code-prettify/blob/master/README.md#usage).

**Comment by jira-importer** [comment]
**[Jeremy Landis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hazendaz)** commented

Sounds like highlight.js is a good replacement.  See https://github.com/highlightjs/highlight.js/

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

This regular expression does not match doxia v2 output which should add the necessary `prettyprint` class: https://github.com/apache/maven-fluido-skin/blob/59b104d843e65eed6023d192b95f875e1e0e2b0a/src/main/resources/META-INF/maven/site.vm#L246.

For Doxia v1 the regular expression from Fluido 1.x matches though: https://github.com/apache/maven-fluido-skin/blob/32835a68dbf391198a4da10582f9908277d24309/src/main/resources/META-INF/maven/site.vm#L251

This is a regression of https://github.com/apache/maven-fluido-skin/commit/c093ebf06ae8d60cb3947dbf0bec3d05d4313c73 (MSKINS-221) [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Why do you expect a div now with both classes `verbatim` and `source`? This does not reflect reality (at least not for this markdown source: https://github.com/awxiaoxian2020/code-render-bug/blob/964ca90272fa95e8c2996e6f439122bce54eb1ef/src/site/markdown/test.md?plain=1#L1 which seems to trigger https://github.com/apache/maven-doxia/blob/d050b2d23e8a3b4318638b05d669d723aebb6743/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/Xhtml5BaseSink.java#L854 without the necessary attribute....)

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

The markdown source

```
```java
int a=1;
```

```

is converted into the following (intermediate) HTML source


```

\<pre>\<code class="language-java">int a=1;
\</code>\</pre>

```

which emits the following Sink events

```

verbatim(no attributes);
inline(attribute for "code");
test("int a=1;");
_inline():
_verbatim();

```

The verbatim calls does not have the mandatory "source" attribute, because at that point in time the parser has not yet seen the following `code` element.
```

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Instead of relying on additional class attributes in `div` I would rather always look for the `code` element in https://github.com/apache/maven-fluido-skin/blob/59b104d843e65eed6023d192b95f875e1e0e2b0a/src/main/resources/META-INF/maven/site.vm#L246 to add the necessary `prettyprint` class to the surrounding `pre`.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Let's slow down here a bit and analyze what is first produced and what is then perfomed by the skin because these are two distinct issues.
What produces Doxia?

```java
    public void verbatim(SinkEventAttributes attributes) {
        if (paragraphFlag) {
            // The content of element type "p" must match
            // "(a|br|span|bdo|object|applet|img|map|iframe|tt|i|b|u|s|strike|big|small|font|basefont|em|strong|
            // dfn|code|q|samp|kbd|var|cite|abbr|acronym|sub|sup|input|select|textarea|label|button|ins|del|script)".
            paragraph_();
        }

        verbatimFlag = true;

        MutableAttributeSet atts = SinkUtils.filterAttributes(attributes, SinkUtils.SINK_VERBATIM_ATTRIBUTES);

        if (atts == null) {
            atts = new SinkEventAttributeSet();
        }

        boolean source = false;

        if (atts.isDefined(SinkEventAttributes.DECORATION)) {
            source = "source"
                    .equals(atts.getAttribute(SinkEventAttributes.DECORATION).toString());
        }

        SinkEventAttributes divAtts = null;
        String divClass = "verbatim";

        if (source) {
            divClass += " source";
        }

        divAtts = new SinkEventAttributeSet(SinkEventAttributes.CLASS.toString(), divClass);

        atts.removeAttribute(SinkEventAttributes.DECORATION);

        writeStartTag(HtmlMarkup.DIV, divAtts);
        writeStartTag(HtmlMarkup.PRE, atts);
    }
```

There is a regular verbatim with `pre` and one attributed with source to make a distinction for source code. Both Xdoc and APT have the difference as well: verbatim (fixed length) and source code. If we want to change that to `<code>` instead of `<pre>` for source code we need to change/fix/adapt Doxia.
What does the Fluido Skin?

```
replaceAll( '<div class="verbatim source">(\r?\n)?<pre>', '<div class="verbatim source"><pre class="' + $sourceStyle + '">' )
```

Exactly replacing that what the skin and pre prettifier expect.

If Markdown, as it obviously is, produces a different output that is a problem with the Markdown sink we need to take care of. The actual skin should not know anything about the source format, IMHO.

Note: Handling with APT:

```java
    public void traverse() throws AptParseException {
        AptParser.this.sink.verbatim(source ? SinkEventAttributeSet.SOURCE : null);
        AptParser.this.sink.text(text);
        AptParser.this.sink.verbatim_();
    }
}
```

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Using custom HTML classes is IMHO counterintuitive and redundant. XHTML has a perfectly valid way to express both block code statements via `<pre><code>` and inline code statements with just `{}code{`}.
Both are also used AFAIK when serializing APT code blocks (i.e. boxed verbatim) statements to XHTML. The additional classes just blow up the HTML unnecessarily (but they don't do any harm either).

 

To rephrase: I would expect that

```java
sink.verbatim(SinkEventAttributeSet.SOURCE);
...
```

and

```java
sink.verbatim(null);
sink.inline(Attribute for "code");
...
```

would emit exactly the same XHTML output

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Regarding:

> If Markdown, as it obviously is, produces a different output that is a problem with the Markdown sink we need to take care of.

This is not only a problem for the Markdown Parser, but also the XHTML Parser. Both don't call sink.verbose(...) with the source attribute.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

OK, let's discuss how to emit proper HTML code in a clean room impl: The `div` with the `pre` is redudant, so are the CSS classes. I agree with you what with `sink.verbatim(null)` a `<pre>` should be printed and with `sink.verbatim(SinkEventAttributeSet.SOURCE)` `<pre><code>`. After that we need to make sure that parsers properly invoke right events than I am totally fine with simplifying Doxia in this regard. As soon as Doxia is fixed Fluido skin can be fixed. Is that OK for you [Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [ab1e6a348671a3413bb51da45f1e4cfa97a3c5da](https://gitbox.apache.org/repos/asf?p=maven-fluido-skin.git;a=commit;h=ab1e6a348671a3413bb51da45f1e4cfa97a3c5da).

**Comment by jira-importer** [comment]
**[Xavi Lee](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER284203)** commented

It seems that we didn't fix this issue at all. Did you confirm with my repro repo? [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)
